platform-sdk/swirlds-state-impl/src/main/java/com/swirlds/state/merkle/disk/OnDiskReadableKVState.java

```diff
@@ -83,24 +88,56 @@ protected V readFromDataSource(@NonNull K key) {
     @Override
     protected Iterator<K> iterateFromDataSource() {
         // Log to transaction state log, what was iterated
-        logMapIterate(getStateKey(), virtualMap, keyCodec);
-        return new OnDiskIterator<>(virtualMap, keyCodec);
+        logMapIterate(getLabel(), megaMap, keyCodec);
+        return new OnDiskIterator<>(megaMap, keyCodec);
     }
 
     /** {@inheritDoc} */
     @Override
     public long size() {
-        final var size = virtualMap.size();
+        final var size = megaMap.size();
         // Log to transaction state log, size of map
-        logMapGetSize(getStateKey(), size);
+        logMapGetSize(getLabel(), size);
         return size;
     }
 
     @Override
     public void warm(@NonNull final K key) {
-        final var kb = keyCodec.toBytes(key);
         // FUTURE work: remove legacy hash code
         final int legacyKeyHashCode = Objects.hash(key); // matches OnDiskKey.hashCode()
-        virtualMap.warm(kb, legacyKeyHashCode);
+        megaMap.warm(getMegaMapKey(key), legacyKeyHashCode);
+    }
+
+    // TODO: test this method
+    // TODO: refactor? (it is duplicated in OnDiskWritableKVState)
+    /**
+     * Generates a key for identifying an entry in the MegaMap data structure.
+     * <p>
+     * The key consists of:
+     * <ul>
+     *   <li>The first 2 bytes: the state ID (unsigned 16-bit, big-endian)</li>
+     *   <li>The remaining bytes: the serialized form of the provided key</li>
+     * </ul>
+     * The state ID must be within [0..65535].
+     * </p>
+     *
+     * @param key the key to serialize and append to the state ID
+     * @return a {@link Bytes} object containing the state ID followed by the serialized key
+     * @throws IllegalArgumentException if the state ID is outside [0..65535]
+     */
+    private Bytes getMegaMapKey(final K key) {
+        final int stateId = getStateId();
+
+        if (stateId < 0 || stateId > 65535) {
+            throw new IllegalArgumentException("State ID " + stateId + " must fit in [0..65535]");
+        }
+
+        final ByteBuffer buffer = ByteBuffer.allocate(2).order(ByteOrder.BIG_ENDIAN);
+        buffer.putShort((short) stateId);
+        final Bytes stateIdBytes = Bytes.wrap(buffer.array());
+
+        final Bytes keyBytes = keyCodec.toBytes(key);
+
+        return stateIdBytes.append(keyBytes);
```

Comment:
It would be too much pressure on GC, too many short living objects are allocated. What about
```java
final byte[] bytes = new byte[Short.BYTES + keyCodec.measureRecord(key)];
final BufferedData writer = BufferedData.wrap(bytes);
writer.writeShort((short) stateId);
keyCodec.write(key, writer);
return Bytes.wrap(bytes);
```

This is still not ideal, but slightly better than above.

Reply:
I agree. Optimized it, there are a couple of details:
1. There is no `writeShort` method -- made a workaround.
2. Wdyt about exception handling?